# Enigma Internship Application

Will Geary (willcgeary@gmail.com)

#### Prompt:

You are given the first 500 rows of a dataset (attached) that was created using tables from the Adverse Events database, released by the U.S. Food & Drug Administration. Imagine that you were asked to pre-process the dataset so that your coworker can perform exploratory analysis on it. Find all the potential problems with the dataset that might affect his or her results. Don't be afraid to mention any issue, even if it seems trivial. We expect you to send an enumerated list with as many data problems as you can spot.

Notes:

* Issues of a similar nature that exist within different fields should be considered separately;

* Describe every data problem in one concise and specific sentence;

* If you suggest a possible explanation for a given issue, please include it at the end of the data issues list;

* Take as much time as you need;

* Use any tool(s) you'd like.

## View the data

In [5]:
%%bash

cat bad_data.csv

primaryid,drug_name,validated_verbatim_drugname,dose_verbatim,patient_gender,patient_weight,weight_unit,report_date
7024584,BIMATOPROST AND TIMOLOL MALEATE,1,"1 GTT, QD",F,63,KG,20101001
7029107,LETROX,2,50MG PER DAY,F,61,KG,20101005
7030380,DEXAMETHASONE,1,20 MG, SINGLE, INTRAVENOUS,M,77,KG,20100930
7031597,BUPROPION HCL,1,100MG BID PO,F,117,LBS,20100930
7046812,SUBUTEX,1,DOSES VARY FROM 8 - 12 MG,F,0,"",20101014
7047487,BIMATOPROST,1,"1 GTT, QD",F,63,KG,20101015
7048917,NASONEX,1,"",M,0,"",20101018
7051833,CHOLESTEROL TABLETS,2,"",F,0,"",20101019
7055438,FLUOXETINE,1,UNK,"",0,"",20101021
7055440,FLUOXETINE,1,UNK,"",0,"",20101021
7056975,SUBUTEX,1,DOSES VARY FROM 8 -
12 MG,F,0,"",20101021
7058702,BEZAFIBRATE,2,400 MG,F,0,"",20101022
7059599,NASONEX,1,"","",0,"",20101025
7064231,AMLODIPINE,1,"10 MG, UNK",M,140,KG,20101027
7064309,ANASTROZOLE,1,1MG DAILY PO,F,194,LBS,20101015
7065351,VALSARTAN,1,UNK,M,140,KG,20101027
7070535,ASPIRIN,2,UNK,"",0,"",20101029
7071923,ALBUTEROL INHALER,2,"",

## Problems with the data (by field)

*Note: row indices use zero-based numbering*

#### primaryid
1) Row 11 is missing primaryid

#### drug_name
2) Some rows have two drug names, i.e. "BIMATOPROST AND TIMOLOL MALEATE"

3) Row 11 is missing drug_name

*Steps to clean this column may include:*
* Seperate each row with two drug names (i.e. has drug_name with the word "and" in it) into two separate rows

#### validated_verbatim_drugname
4) Row 11 is missing validated_verbatim_drugname

#### dose_verbatim
5) Values have inconsistent unit naming conventions, including:
* "50MG PER DAY"
* "DOSES VARY FROM 8 - 12 MG"
* "1 TABLET 1 TIMES PER 1 DAY"
* "80 MG ORAL"
* "AS-NEEDED BASIS"

6) Many rows are missing dose_verbatim values

7) It would take sigificant work to clean up the inconsistent dose_verbatim naming conventions


*Steps to clean this column may include:*
* Separate amount and frequency into separate columns with consistent units, such as dose_amount = 50 (i.e 50 MG) and dose_freq = 1 (i.e. 1x per day)
* Create a system for categorizing dose frequencies. Some doses are once or twice per day, others are "as-needed" and yet others are weekly. One could numerate all of the various frequency types into key-value pairs, such as {1="1x per day", 2 = "2x per day", 3 = "1x per week", 4 = "2x per week", 5 = "as needed"}
* Some observations have varying doses, such as "doses vary from 8 - 12 mg". To account for this, one could create optional minimum and maximum fields to contain doses that have ranges.
* Is information on the dose amounts and frequencies material to the analysis that will be performed on this data? If it is not material, it may be preferable to simply drop the field altogether rather than spending a large amount of time cleaning this data.

#### patient_gender
8) Many rows are missing patient_gender values

9) Some rows have what appears to be the patient_gender value in the wrong column (row 2, row 106)

Steps to clean this column may include:
* Encode missing patient_gender values as Unknown
* Fix rows 2 and 106 such that patient_gender values are in the correct column

#### patient_weight
10) Many rows have patient_weight missing or zero

11) Some rows have what appears to be the patient_weight value in the wrong column (row 2, row 106)

*Steps to clean this column might include:*
* Replace missing or zero patient_weight's with "NA" so that values of zero aren't accidentally taken into account (i.e. we wouldn't want a zero in this column to be included when calculating average patient_weight)
* Fix rows 2 and 106 such that patient_weight values are in the correct column
    
#### weight_unit
12) Many rows have missing weight_unit

13) Most rows use metric system, but some use imperial system

14) Some rows have what appears to be the weight_unit value in the wrong column (row 2, row 106)

*Steps to clean this column might include:*
* Convert all imperial system weight_units to metric system
* Fix rows 2 and 106 such that patient_gender values are in the correct column
    
#### report_date
15) The report_dates are in different date formats, including: "20100930", "2010-09-30"

16) Several report_dates are missing the day. For example: BYETTA = 201011, SYMBICORT = 201011

17) Some rows have what appears to be the patient_weight value in the wrong column (row 2, row 106)

*Steps to clean this column might include:*
* Use the python datetime module to convert report_dates into datetime objects
* If format of report_date == "20100930", then use datetime.strptime("YYYMMDD")
* If format of report_date == "2010-09-30", then use datetime.strptime("YYYY-MM-DD")

## Problems with specific rows

18) Row 2
* dose_verbatim is split into three columns. This is causing patient_gender, patient_weight, weight_unit and report_date columns to be pushed to the right
    
19) Row 10
* dose_verbatim is being split into two rows. This is causing patient_gender, patient_weight, weight_unit, report_date values to be pushed downwards one row
    
20) Row 106
* dose_verbatim is split into three columns. This is causing patient_gender, patient_weight, weight_unit and report_date columns to be pushed to the right
    
21) Row 501 (last row)
* the last row is missing the following fields: dose_verbatim, patient_gender, patient_weight, weight_unit, report_date